File: spring/src/test/java/com/example/demo/service/OrderServiceTest.java
//package com.example.demo.service;
//
//import com.example.demo.config.annotation.LocalBootTest;
//import com.example.demo.entity.order.Order;
//import com.example.demo.entity.product.Product;
//import com.example.demo.entity.user.User;
//import org.junit.jupiter.api.BeforeAll;
//import org.junit.jupiter.api.DisplayName;
//import org.junit.jupiter.api.Test;
//import org.springframework.beans.factory.annotation.Autowired;
//import org.springframework.security.test.context.support.WithUserDetails;
//import org.springframework.transaction.annotation.Transactional;
//
//import java.util.Arrays;
//import java.util.List;
//import java.util.stream.Collectors;
//
//import static org.junit.jupiter.api.Assertions.assertEquals;
//import static org.junit.jupiter.api.Assertions.assertTrue;
//
//@LocalBootTest
//@Transactional
//class OrderServiceTest {
//
//    @Autowired
//    private OrderService orderService;
//
//    @Autowired
//    private ProductService productService;
//
//    @Autowired
//    private UserService userService;
//
//    private User testUser;
//    private List<Product> products;
//
//    @BeforeAll
//    void setUp() {
//        this.testUser = User.builder()
//                .loginId("demo")
//                .email("email@demo.com")
//                .savePassword("password")
//                .build();
//        products = Arrays.stream(
//                new Product[]{
//                        Product.builder().name("상품이름 1").price(10000).build(),
//                        Product.builder().name("상품이름 2").price(20000).build(),
//                        Product.builder().name("상품이름 3").price(30000).build()
//                }).map(product -> productService.save(product))
//                .collect(Collectors.toList());
//    }
//
//    @Test
//    @DisplayName("상품 주문 테스트")
//    @WithUserDetails("demo")
//    void saveUserTest() throws Exception {
//        // given
//        List<Long> mockProductIds = products.stream().map(Product::getId).collect(Collectors.toList());
//        Order order = orderService.orderProducts(mockProductIds);
//
//        // when
//        List<Order> results = orderService.findAll();
//        Order newOrder = results.stream().filter(orderByUser -> orderByUser.getOrderUser().getId().equals(testUser.getId())).findFirst().get();
//        List<Long> newOrderProductIds = newOrder.getOrderList().stream().map(o -> o.getProduct().getId()).collect(Collectors.toList());
//
//        // then
//        assertEquals(order.getId(), newOrder.getId());
//        assertTrue(mockProductIds.containsAll(newOrderProductIds));
//    }
//}

File: spring/src/test/java/com/example/demo/service/OrderServiceTest.kt
package com.example.demo.service

import com.example.demo.config.annotation.JwtLoginUser
import com.example.demo.config.annotation.LocalBootTest
import com.example.demo.product.domain.Product
import com.example.demo.user.domain.User
import com.example.demo.generator.MockUserBuilder
import com.example.demo.order.service.OrderService
import com.example.demo.product.domain.ProductRepository
import com.example.demo.product.service.ProductService
import com.example.demo.user.domain.UserRepository
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.transaction.annotation.Transactional

@LocalBootTest
@Transactional
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class OrderServiceTest {
    @Autowired
    lateinit var orderService: OrderService

    @Autowired
    lateinit var productService: ProductService

    @Autowired
    lateinit var userRepository: UserRepository

    @Autowired
    lateinit var productRepository: ProductRepository

    @Autowired
    lateinit var mockUserBuilder: MockUserBuilder

    private val rawPassword = "password"
    private var testUser: User? = null
    private var products: List<Product>? = null

    @BeforeAll
    @Throws(Exception::class)
    fun setUp() {
        testUser = mockUserBuilder.build("demo", "email@demo.com", rawPassword)
        products = listOf(
            Product(name = "상품이름 1", price = 10000),
            Product(name = "상품이름 2", price = 20000),
            Product(name = "상품이름 3", price = 30000)
        ).map {
            productService.save(it)
        }
    }

    @Test
    @DisplayName("상품 주문 테스트")
    @JwtLoginUser(loginId = "demo")
    fun orderProductTest() {
        // given
        val mockProductIds = products!!.map { it.id ?: fail("mock 상품 저장 실패") }
        val order = orderService.orderProducts(mockProductIds)

        // when
        val results = orderService.findAllOrderByLoginUser()
        val newOrder = results.first { it.orderUser.id == testUser?.id }
        val newOrderProductIds = newOrder.orderList.map { it.product.id }

        // then
        assertNotNull(order.id)
        assertEquals(order.id, newOrder.id)
        Assertions.assertTrue(mockProductIds.containsAll(newOrderProductIds))
    }

    @AfterAll
    fun teardown() {
        productRepository.deleteAllById(products!!.map{ it.id })
        userRepository.delete(testUser!!)
    }
}
